File: ch4/traversal/UVa11838.cpp
// Come and Go
// check if the graph is strongly connected, i.e. the SCC of the graph is the graph itself (only 1 SCC)

#include <bits/stdc++.h>
using namespace std;

typedef pair<int, int> ii;
typedef vector<int> vi;
typedef vector<ii> vii;

enum { UNVISITED = -1 };

int dfsNumberCounter, numSCC;
vector<vii> AL, AL_T;
vi dfs_num, dfs_low, S, visited;                 // global variables
stack<int> St;

void tarjanSCC(int u) {
  dfs_low[u] = dfs_num[u] = dfsNumberCounter;    // dfs_low[u]<=dfs_num[u]
  dfsNumberCounter++;                            // increase counter
  St.push(u);                                    // remember the order
  visited[u] = 1;
  for (auto &[v, w] : AL[u]) {
    if (dfs_num[v] == UNVISITED)
      tarjanSCC(v);
    if (visited[v])                              // condition for update
      dfs_low[u] = min(dfs_low[u], dfs_low[v]);
  }

  if (dfs_low[u] == dfs_num[u]) {                // a root/start of an SCC
    ++numSCC;                                    // when recursion unwinds
    while (1) {
      int v = St.top(); St.pop(); visited[v] = 0;
      if (u == v) break;
    }
  }
}

void Kosaraju(int u, int pass) { // pass = 1 (original), 2 (transpose)
  dfs_num[u] = 1;
  vii &neighbor = (pass == 1) ? AL[u] : AL_T[u]; // by ref to avoid copying
  for (auto &[v, w] : neighbor)                  // C++17 style, w ignored
    if (dfs_num[v] == UNVISITED)
      Kosaraju(v, pass);
  S.push_back(u); // as in finding topological order in Section 4.2.5
}

int main() {
  int N, M;
  while (scanf("%d %d", &N, &M), (N || M)) {
    AL.assign(N, vii());
    AL_T.assign(N, vii()); // the transposed graph
    while (M--) {
      int V, W, P; scanf("%d %d %d", &V, &W, &P); --V; --W; // to 0-based indexing
      AL[V].emplace_back(W, 1); // always
      AL_T[W].emplace_back(V, 1);
      if (P == 2) { // if this is two way, add the reverse direction
        AL[W].emplace_back(V, 1);
        AL_T[V].emplace_back(W, 1);
      }
    }

    // run Tarjan's SCC code here
    dfs_num.assign(N, UNVISITED); dfs_low.assign(N, 0); visited.assign(N, 0);
    while (!St.empty()) St.pop();
    dfsNumberCounter = numSCC = 0;
    for (int u = 0; u < N; ++u)
     if (dfs_num[u] == UNVISITED)
       tarjanSCC(u);

    // // run Kosaraju's SCC code here
    // S.clear(); // first pass: record the post-order of original graph
    // dfs_num.assign(N, UNVISITED);
    // for (int u = 0; u < N; ++u)
    //   if (dfs_num[u] == UNVISITED)
    //     Kosaraju(u, 1);
    // int numSCC = 0; // second pass: explore SCCs using first pass order
    // dfs_num.assign(N, UNVISITED);
    // for (int i = N-1; i >= 0; --i)
    //   if (dfs_num[S[i]] == UNVISITED)
    //     numSCC++, Kosaraju(S[i], 2);             // on transposed graph

    // if SCC is only 1, print 1, otherwise, print 0
    printf("%d\n", numSCC == 1 ? 1 : 0);
  }
  return 0;
}

File: ch4/traversal/UVa11838.ml
open Scanf
open Printf

let range n = Array.init n (fun x -> x)

type graph = int list array (* Adjacency List representation *)

(** Returns a list of all strongly connected components of the directed graph *)
let tarjan_scc (adj : graph) : int list list =
  let n = Array.length adj in
  let low = Array.make n 0 in
  let num = Array.make n (-1) in
  let counter = ref 0 in
  let stack = ref [] in
  let on_stack = Array.make n false in
  let sccs = ref [] in
  let rec visit u =
    low.(u) <- !counter;
    num.(u) <- !counter;
    incr counter;
    stack := u :: !stack;
    on_stack.(u) <- true;
    adj.(u) |> List.iter (fun v ->
      if num.(v) = -1 then begin
        visit v;
        low.(u) <- min low.(u) low.(v)
      end
      else if on_stack.(v) then
        low.(u) <- min low.(u) num.(v)
    );
    if low.(u) = num.(u) then begin
      let rec pop_scc acc = function
        | hd :: tl ->
          if hd = u then hd :: acc, tl
          else pop_scc (hd :: acc) tl
        | [] -> assert false
      in
      let scc, stack' = pop_scc [] !stack in
      stack := stack';
      scc |> List.iter (fun v -> on_stack.(v) <- false);
      sccs := scc :: !sccs
    end
  in
  range n |> Array.iter (fun u -> if num.(u) = -1 then visit u);
  !sccs

let kosaraju adj_list adj_list_t =
  let n = Array.length adj_list in
  let visited = Array.make n false in

  let rec dfs adj order u =
    visited.(u) <- true;
    u :: (adj.(u) |> List.fold_left (fun acc v ->
        if visited.(v) then acc else dfs adj acc v
      ) order) in
  
  let order =
    range n |> Array.fold_left (fun acc u ->
      if visited.(u) then acc else dfs adj_list acc u ) [] in
  
  Array.fill visited 0 n false;
  order |> List.fold_left (fun sccs u ->
    if visited.(u) then sccs else dfs adj_list_t [] u :: sccs ) []

let solve n m =
  let adj = Array.make n [] in
  let adj' = Array.make n [] in
  for _ = 1 to m do
    let u, v, t = scanf "%d %d %d\n" (fun x y z -> (x - 1, y - 1, z)) in
    adj.(u) <- v :: adj.(u);
    adj'.(v) <- u :: adj'.(v);
    if t = 2 then begin (* if this is two way, add the reverse direction *)
      adj.(v) <- u :: adj.(v);
      adj'.(u) <- v :: adj'.(u);
    end
  done;
  let sccs = kosaraju adj adj' in
  if List.length sccs = 1 then printf "1\n"
  else printf "0\n"

let () =
  let rec loop () =
    let n, m = scanf "%d %d\n" (fun x y -> (x, y)) in
    if n = 0 && m = 0 then ()
    else begin
      solve n m;
      loop ()
    end in
  loop ()
